Migrated `AddCategoryFragment` to ViewBinding

neeldoshii opened the pull request:
## Description 

This pull request migrates the `AddCategoryFragment`  to use ViewBinding, improving type safety and eliminating the need for findViewById calls. 

## Fixes 
* #19180

## Screenshots/ Video

https://github.com/wordpress-mobile/WordPress-Android/assets/60827173/7914585c-6459-4c63-9b67-d725fb97ea70




## Steps to reproduce 
1. Create a Post
2. Click on Any of your Posts
3. Go to Post Settings
4. Click on Categories
5. On top right click on Add Button (+ Button) for adding new categories

## PR Submission Checklist:

- [ ] I have completed the Regression Notes.
- [X] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Testing Checklist (strike-out the not-applying and unnecessary ones):

- [X] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [X] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

neeldoshii committed:
Migrated `AddCategoryFragment` to ViewBinding

neeldoshii commented:
Keeping each class as a seperate commit for ease in revert. If not needed then will rebase it.

ParaskP7 reviewed (commented):
👋 @neeldoshii and thanks so much for submitting this PR!

You shared a screenshot/video with steps to reproduce and test this migration, many thanks for that, but:
  1. Unfortunately you actually tested a different workflow, the `PrepublishingAddCategoryFragment.kt` functionality, and not the `AddCategoryFragment`. 
  2. This means that you actually didn't verify that your change didn't brake anything. Which, might result in us merging this and then getting into problems.

I actually found how to test this `AddCategoryFragment` myself, but, I would like you to invest some extra time into doing that yourself too. Thus, I am not going to give you the answer here. I don't do that just to make it more difficult for you. Instead, I do that to help/guide you into being able to find it yourself, not just for this specific case, but for all the contributions you will be adding to this repo going forward.

FYI: What I did to figure that out is using the debugger. I added breakpoints to classes where I thought they might trigger this workflow. And then, after some time, I found where `AddCategoryFragment` is actually being used. I would like you to do the same.

-----

As I have already explained, only with property testing, testing all edge-cases actually, will we make sure that we are not adding any regression to this screen. Thus me insisting on properly testing this before and after the change.

I hope all that makes sense and I am not overwhelming you with too much info. 🙏

ParaskP7 commented:
Also, since we need to run this PR on CI, I need you to fix all the checks first, otherwise I am not able to push this branch and trigger our CI to run on it. In this case, `CheckStyle` fails with the below:

```
> Task :WordPress:checkstyle
[ant:checkstyle] [ERROR] /Users/.../WordPress-Android/WordPress/src/main/java/org/wordpress/android/ui/posts/AddCategoryFragment.java:31:32: Name 'binding' must match pattern '^m[a-zA-Z0-9]*$'. [MemberName]
[ant:checkstyle] [ERROR] /Users/.../WordPress-Android/WordPress/src/main/java/org/wordpress/android/ui/posts/AddCategoryFragment.java:55:55: ',' is not followed by whitespace. [WhitespaceAfter]
[ant:checkstyle] [ERROR] /Users/.../WordPress-Android/WordPress/src/main/java/org/wordpress/android/ui/posts/AddCategoryFragment.java:55:60: ',' is not followed by whitespace. [WhitespaceAfter]
```

To run `CheckStyle` locally use this command: `./gradlew checkstyle`

PS: I also suggest that you run all checks locally before pushing and opening any PR, just to make sure our CI will also succeed afterwards (see [pipeline.yml](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/.buildkite/pipeline.yml)). For you, the most important checks and tests command we use on CI are the following:
1. `./gradlew checkstyle` for Java classes
2. `./gradlew detekt` for Kotlin classes
3. `./gradlew lintWordpressJalapenoDebug` + `./gradlew lintJetpackJalapenoDebug` for Android Lint
4. `./gradlew testWordpressJalapenoDebugUnitTest` + `./gradlew testJetpackJalapenoDebugUnitTest` for Unit Tests

I hope the above will help you... 🙏 

-----

I am going to pause reviewing #20941 until we are done with this PR first, and maybe then, after you apply what you learned here on this other PR, I'll go for another round of review on it. I hope that's okay with you. 😊

neeldoshii commented:
Thank you @ParaskP7 for reviewing 🙏. 

> Unfortunately you actually tested a different workflow, the PrepublishingAddCategoryFragment.kt functionality, and not the AddCategoryFragment.

Thank you for spotting this mistake. Updated the steps to reproduce the dialog in the description along with the video. 

> I actually found how to test this AddCategoryFragment myself, but, I would like you to invest some extra time into doing that yourself too. Thus, I am not going to give you the answer here. I don't do that just to make it more difficult for you. Instead, I do that to help/guide you into being able to find it yourself, not just for this specific case, but for all the contributions you will be adding to this repo going forward.

Thank you providing a good learning curve to me 🚀

> FYI: What I did to figure that out is using the debugger. I added breakpoints to classes where I thought they might trigger this workflow. And then, after some time, I found where AddCategoryFragment is actually being used. I would like you to do the same.

Thanks for this, this is how I reproduced. Created two debugger breakpoint one at the class and second when the dialog is initialized/created to know when the dialog is called. Then kept playing in the app to wait when the breakpoint gets triggered 😅. This way It ensured this is the correct place.

--

Made the changes so the test passes and committed the new changes. ✅

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/posts/AddCategoryFragment.java:
```diff
@@ -136,4 +130,10 @@ public void onSaveInstanceState(Bundle outState) {
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        mBinding = null; // ensures binding reference gets null on fragment destroy
```
Minor (🔍): This comment is not too necessary as it is self-explanatory what is happening here and why. No need for you to do anything at this point, this is just an FYI for you for your next PRs.

neeldoshii replied:
Updated the code.

ParaskP7 replied:
Great, thanks for that @neeldoshii , although, as I told you this was totally optional and you could have avoided it, I added this comment for educational purposes only! 🥇 

FYI: However, I want to emphasise here again on avoiding to rebase + force push on any PR ([context](https://github.com/wordpress-mobile/WordPress-Android/pull/20941#discussion_r1636312985)).

neeldoshii replied:
Sure 👍.

ParaskP7 reviewed (approved):
👋 @neeldoshii !

Thank you so much for your contribution to JP/WPAndroid with this PR! 🥇 

I have reviewed and tested this PR as per the instructions, along with triggering CI on it via a [draft](https://github.com/wordpress-mobile/WordPress-Android/pull/20970) PR, which I just pushed on the main repo, everything works as expected, good job! 🌟 

-----

> Thank you for spotting this mistake. Updated the steps to reproduce the dialog in the description along with the video.

Thanks for updating the steps to reproduce, much appreciated! 🥇 

> Thank you providing a good learning curve to me 🚀

Thank YOU! 🥇 

> Thanks for this, this is how I reproduced. Created two debugger breakpoint one at the class and second when the dialog is initialized/created to know when the dialog is called. Then kept playing in the app to wait when the breakpoint gets triggered 😅. This way It ensured this is the correct place.

👍 

> Made the changes so the test passes and committed the new changes. ✅

🚀

ParaskP7 commented:
🎉 @neeldoshii , your first contribution to JP/WPAndroid is now history and will be live with the next release, thanks so much for this work! 🥇
